### PR TITLE
Topology aware msg delivery 1301

### PR DESCRIPTION
**Pull request by zoemccormick**
Screenshot 2023-10-24 at 12 59 28 PM

Screenshot 2023-11-17 at 2 12 24 PM

### CODE REVIEW

**Review by Ulminator — APPROVED** [review]
lgtm 👍

File: nsqadmin/static/js/views/channel.hbs [hunk 1]
```diff
@@ -176,8 +176,12 @@
             <tr>
                 <th>Client Host</th>
                 <th>User-Agent</th>
+                <th>Client Region</th>
+                <th>Client Zone</th>
                 <th>Attributes</th>
```
Comment (jehiah) [inline nsqadmin/static/js/views/channel.hbs]: style/ux comments:

In general i feel this makes the table too busy.

I think these would be better added to the Attributes field as a label (w/ a specific coloring for the zone + region value). I think the host zone/region should similarly be labels in the `nsqd host` cell.

The think you should be able to get at a glance is when a connection is a zone-local match, or when a connection is a region-local match. I would almost be ok not showing the nsqd host zone/region if we added that. One option -we could use a coloring of the background on the row to flag zone and region local? should we add region + zone in grey but color them when they are a zone-local or region-local match so they pop?

File: internal/clusterinfo/data.go [hunk 2]
```diff
@@ -352,6 +352,8 @@ func (c *ClusterInfo) GetNSQDProducers(nsqdHTTPAddrs []string) (Producers, error
 		Hostname         string `json:"hostname"`
 		HTTPPort         int    `json:"http_port"`
 		TCPPort          int    `json:"tcp_port"`
+		TopologyZone     string `json:"topology_zone"`
+		TopologyRegion   string `json:"topology_region"`
```
Comment (jehiah) [inline internal/clusterinfo/data.go]: Should most of these spots should be `omitempty`?

**Review by jehiah — COMMENTED** [review]
Overall LGTM

File: nsqadmin/static/js/views/channel.hbs (hunk 1 above)
Comment (zoemccormick) [reply nsqadmin/static/js/views/channel.hbs]: haha, we also originally though it would be good to do a colored matching, i will look back into this

Comment (jehiah) [reply nsqadmin/static/js/views/channel.hbs]: also to clarify the framing of my comments - the existing state is fine to land as a starting point.

**Comment by mreiferson** [comment]
is the idea to land this in @jehiah's fork and then open a PR upstream with a larger set of changes?

**Comment by jehiah** [comment]
@mreiferson yup - plan is to work off branches on a fork to get it end-to-end validated on the Bitly stack, then to loop back to upstream after that (probably some work on that in Dec, but most likely January); in the mean time any comments/suggestions/feedback welcome.

edit: sorry if my phrasing "fine to land as a starting point" is confusing in that context

**Comment by zoemccormick** [comment]
style updates:

Screenshot 2023-11-17 at 2 04 01 PM

or with default label style:

Screenshot 2023-11-17 at 2 12 24 PM

**Pull request merged**